fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/PrefetchableBufferedChecksumIndexInput.java

```diff
@@ -0,0 +1,108 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import com.apple.foundationdb.record.lucene.directory.FDBIndexInput;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+
+import java.io.IOException;
+
+/**
+ * An Index Input that attempts to keep a 160K buffer in front of the read at a minimum.  The goal for this
+ * class is to provide some parallelism for fetches from FDB when we are attempting to merge segments.  The signaling
+ * of when to prefetch occurs in the FDBIndexOutput#copyBytes method.
+ *
+ */
+public class PrefetchableBufferedChecksumIndexInput extends BufferedChecksumIndexInput {
+    private IndexInput main;
+    private static final long PREFETCH_MINIMUM_BUFFER = 16 * 1024 * 10; // Attempt to keep a large enough buffer
+    private long numberOfBytesToPrefetch;
+    private long currentPrefetch = 0;
+    private long totalPrefetch = 0;
+
+    public PrefetchableBufferedChecksumIndexInput(final IndexInput main) {
+        super(main);
+        this.main = main;
+    }
+
+    public boolean supportsPrefetch() {
+        return main instanceof FDBIndexInput;
+    }
+
+    public void setNumberOfBytesToPrefetch(long numberOfBytesToPrefetch) {
+        this.numberOfBytesToPrefetch = numberOfBytesToPrefetch;
+        this.currentPrefetch = 0;
+        this.totalPrefetch = 0;
+        checkBuffer();
+    }
+
+    public void checkBuffer() {
+        if (numberOfBytesToPrefetch > 0) {
+            if (currentPrefetch < PREFETCH_MINIMUM_BUFFER && currentPrefetch < numberOfBytesToPrefetch) {
+                long bytesToFetch = Math.min(PREFETCH_MINIMUM_BUFFER, numberOfBytesToPrefetch);
+                prefetch(bytesToFetch, totalPrefetch);
+                currentPrefetch += bytesToFetch;
+                totalPrefetch += bytesToFetch;
+            }
+        }
+
+    }
+
+    public void prefetch(long numberOfBytes, long currentPrefetch) {
+        if (supportsPrefetch()) {
+            ((FDBIndexInput)main).prefetch(numberOfBytes, currentPrefetch);
+        }
+    }
+
+    @Override
+    public byte readByte() throws IOException {
+        decrement();
+        checkBuffer();
+        return super.readByte();
+    }
+
+    @Override
+    public void readBytes(final byte[] b, final int offset, final int len) throws IOException {
+        decrement(len);
+        checkBuffer();
+        super.readBytes(b, offset, len);
+    }
+
+    @Override
+    public void seek(final long pos) throws IOException {
+        decrement(pos - getFilePointer());
```

Comment:
If seeking past the `currentPrefetch` window, this will cause it to go negative. Maybe that's okay, but the effect will them be that each call to `checkBuffer` will advance the end until we catch up. This is, it will prefetch everything that was seeked over. Is that what we want?

Reply:
Ok, let me change.